documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -0,0 +1,610 @@
+    is complete, logging the progress as it goes
+- Use
+  [`terminate_instance_in_auto_scaling_group`](https://boto3.amazonaws.com/v1/documentation/api/1.26.86/reference/services/autoscaling/client/terminate_instance_in_auto_scaling_group.html)
+  to terminate the instance. Make sure to set `ShouldDecrementDesiredCapacity`
+  to `True` to ensure that the ASG does not try to replace the instance. This
+  task should use the
+  [`NONE_SKIPPED` TriggerRule](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#trigger-rules)
+  to ensure that the instances are terminated, even if there are upstream
+  failures. (Skips in local env.)
+- Finally, after all tasks have finished (regardless of success/failure), we
+  should have a cleanup task that calls `set_desired_capacity` to 0. Generally
+  this should be a no-op, but if an instance crashes during reindexing (rather
+  than simply failing during reindexing) the ASG will spin up a replacement and
+  Airflow will not automatically clean it up. This task ensures that any
+  dangling instances are terminated.
+
+```{note}
+It is not possible to retry a single indexer worker with this set up, because once a worker fails the instance is actually terminated (rather than simply stopped). If a task that triggers a reindex is cleared after an instance has been terminated, it will simply fail. The entire reindex must be restarted from the first step in this task group.
+
+However, there is a valuable tradoff to this approach: it ensures that all of the indexer workers in a data refresh are identical, while still allowing us to avoid manual deployments every time the indexer logic changes. For example, imagine some changes to the reindexing logic are merged to `main` while a data refresh is actively underway, and a new Docker image is published. If one indexer worker failed, and it were possible to retry **just** that indexer worker, it would use the new Docker image -- leading to inconsistency in the behavior of different workers within a single data refresh.
+```
+
+### Create the Terraform and Ansible resources needed to deploy the new indexer workers
+
+In this step we will add the resources needed to actually configure the ASGs.
+
+Some important notes:
+
+- Currently, the staging and production workers are split into separate
+  environments (i.e., a staging deploy is used to deploy the 2 staging workers
+  separately from the 6 production workers). It is more accurate to view all 8
+  workers as production instances (i.e., part of the production catalog
+  deployment), which merely _operate_ on different environments. As such all 8
+  should be part of the production deployment, but two separate ASGs which are
+  tagged indicating their intended environment.
+- The playbooks must be updated to check if any of the four _new_ data refresh
+  DAGs are running before deploying, as well.
+- The `user_data` script should be updated to pull the Docker image with the
+  `latest` tag.
```

Comment:
Just a nit to note that this should change if we use the AMI approach, which I think we should, and I aim to have resolved so as not to block the infrastructure work of this IP. It's a conceptual/foundational piece of our ASG-based deployment puzzle that needs to be solved one way or another. It doesn't _need_ to block this project, so you don't need to change from this described approach here using docker to pull the latest image, and the AMI approach builds on that. But just in case we manage to sort it out before we get to this point of the plan, it's worth noting the difference.